nomad/state/node_events.go

```diff
@@ -0,0 +1,175 @@
+package state
+
+import (
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/nomad/nomad/event"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// NodeEvent represents a NodeEvent change on a given Node.
+type NodeEvent struct {
+	Message string
+	NodeID  string
+}
+
+// NNodeDrainEvent holds information related to a Node Drain
+type NodeDrainEvent struct {
+	NodeID        string
+	Allocs        []string
+	DrainStrategy structs.DrainStrategy
+	Message       string
+}
+
+func (s *StateStore) NodeEventsFromChanges(tx ReadTxn, changes Changes) ([]event.Event, error) {
+	var events []event.Event
+
+	var nodeChanges map[string]*memdb.Change
+
+	markNode := func(node string, nodeChange *memdb.Change) {
+		if nodeChanges == nil {
+			nodeChanges = make(map[string]*memdb.Change)
+		}
+		ch := nodeChanges[node]
+		if ch == nil {
```

Comment:
It looks like we're throwing away the second change to a node, why don't we want an event for that too?

Reply:
We do, I'll need to go back and edit this, most of this file is likely going to change but wanted to exercise the transaction changes bits with what node event sourcing may look like